doc/rofilib/configuration.rst

```diff
@@ -84,19 +125,205 @@ Functions
+
+    {
+        "id" : 66,
+        "components" : [ < array of components > ],
+        "joints"     : [ < array of joints >     ]
+    }
+
+where the `component` has three possible values
+
+.. code-block:: json
+
+    [
+        {
+            "type" : "roficom"
+        },
+        {
+            "type" : "UM shoe"
+        },
+        {
+            "type" : "UM body"
+        }
+    ]
+
+and `joint` is represented as
+
+.. code-block:: json
+
+    {
+        "from" : < component >,
+        "destination" : < component >,
+        "sourceToDestination" : < matrix >,
+        "joint" : < joint >
+    }
+
+where possible values of the `joint` are either `RigidJoint` represented as
+
+.. code-block:: json
+
+    {
+        "type" : "rigid"
+    }
+
+or the `RotationJoint` which requires appropriate matrices
+
+.. code-block:: json
+
+    {
+        "type" : "rotational",
+        "axis" : < 4-dimensional array >,
+        "preMatrix"  : < matrix >,
+        "postMatrix" : < matrix >,
+        "min" : < lower-limit – number >,
+        "max" : < upper-limit – number >
+    }
+
+Matrices are, as shown above, represented by 4x4 dimensional array. Or, for
+the identity matrix, you can use a string representation, just write `"identity"`
+instead of `[[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1]]`.
 
-Ideally also include the grammar for configuration.
 
 .. doxygenfunction:: rofi::configuration::serialization::toJSON( const Rofibot& bot, Callback attrCb )
     :project: configuration
 
+The callback is optional. It provides you with the ability to extend the json representation with
+an `"attributes"` property, which can be added to any object within the `json`. It can contain some
+metadata you might use when working with and sharing the configuration description. For details, see
+the section below.
+
 .. doxygenfunction:: rofi::configuration::serialization::toJSON( const Rofibot& bot )
     :project: configuration
 
 .. doxygenfunction:: rofi::configuration::serialization::fromJSON( const nlohmann::json& j, Callback attrCb )
     :project: configuration
 
+Here you can provide a callback function, that is used for parsing the optional `"attributes"` field. If no
+callback is provided, the field, if present, is ignored. The callback is written in the same way as for `toJSON`.
+
 .. doxygenfunction:: rofi::configuration::serialization::fromJSON( const nlohmann::json& j )
     :project: configuration
+
+Attributes callback
+-------------------
+
+You can extend the `json` description of a configuration with `"attributes"` field. This field can be present
+in any object within the configuration, so the callback function has to be able to accept every corresponding
+type. The possible callback for `toJSON` that stores a `ModuleId` to `"attributes"` looks like
+
+.. code-block:: cpp
+
+    overload{ []( const Module& m ) { return nlohmann::json( m.getId() ); },
+              []( const ComponentJoint&, int ) { return nlohmann::json{}; },
```

Comment:
Can you name the arguments? It's not clear from anywhere what the `int` is.

Reply:
Done. I moved the description from functions hidden in `details` to `toJSON` and `fromJSON`, that are exposed.